Dosdebug 47

andrewbird opened the pull request:
I'm experimenting with replacing the trace over ring buffer with a one shot breakpoint. For now I'm just testing with real mode. I have some of it working but:
- [x] Not quite sure where to clean up the one shot breakpoint after it's been hit.
- [x] I must have broken the normal trace code, because after I step over the interrupt using the breakpoint method tracing with just 't' / 'ti' the instruction pointer never moves on.
- [ ] Need to work out why 't' doesn't advance cs:ip in DPMI mode.
- [ ] Need to work out why I have a crash on tracing in DPMI.

stsp commented:
> Not quite sure where to clean up the one shot breakpoint after it's been hit.

Probably immediately?

> breakpoint method tracing with just 't' / 'ti' the instruction pointer never moves on e.g.

Because you didn't remove the BP immediately?

andrewbird commented:
I've now added the one shot cleanup to the bpchk() function but although it does remove the breakpoint I still don't move on. Tonight I'll check I didn't mess up with the refactoring commit before I added in the new stuff.

andrewbird commented:
I found the breakage. Now seems to work for real mode, but I get a crash when in DPMI. I am trying to use the same tracing code for both real and DPMI mode, and that may not be suitable.

stsp commented on src/plugin/debugger/mhpdbgc.c:
```diff
@@ -2415,20 +2362,30 @@ static int bpchk(unsigned int addr)
       trapped_bp_ = i;
       trapped_bp = -2;
+      if (mhpdbgc.brktab[i].is_one_shot) {
+        uint8_t opcode = READ_BYTE(mhpdbgc.brktab[i].brkaddr);
```
Maybe just call the func to clear break-point?

andrewbird replied:
The existing function `mhp_clearbp()` doesn't take an index, only the addr, so it would need to iterate through again. I could make a new function that cleared by index and both could use that if necessary.

stsp replied:
I suspect, even just switching `mhp_clearbp()`
to use index instead of addr is possible, as there
is always index in `trapped_bp`.

andrewbird replied:
I'm just switching `mhp_clearbp()` to take an index, but in order to do so I need to use `trapped_bp` from mhpdbg.c and it's static in mhpdbgc.c. So would you prefer either making it non static, or create a new field in mhpdgbc structure?

stsp replied:
Probably field is better, see what suits better. :)

andrewbird replied:
Actually whilst switching `mhp_clearbp()` to take an index is great for my new stuff I'm finding `trapped_bp` is not available for bpload tidy up, primarily because bpload doesn't use `bkchk()` to test the breakpoint but does it itself. I suspect that bpload could be implemented by using the one_shot breakpoint mechanism now instead of tidying up at program exit, but I'm not looking to make this PR any bigger. So what's best, iterate in bpload tidyup to find the index, or revert to present `mhp_clearbp()` with addr and a new separate function to clear bp by index for my new stuff?

stsp commented:
What kind of crash?
Do normal break-points work fine
in DPMI for you?

andrewbird commented:
>  What kind of crash?
Do normal break-points work fine
in DPMI for you?

Not sure, as I'm new to this DPMI stuff. I just tried something like @ecm-pushbx did in #1178 but with a DJGPP compiled binary and got this in boot.log.

~~~
INT21 at ff10:01e8: AX=6201, BX=00e6, CX=0800, DX=07bf, DS=04bf, ES=c33e        
ERROR: OUCH! attempt to execute interrupt table - quickly dying                 
leavedos(do_int:2590|0) called - shutting down                                  
leavedos thread started                                                         
leavedos thread ended                                                           
ERROR: mhp_poll_loop() reentered                                                
ERROR: mhp_poll_loop() reentered                                                
coopthreads stopped                                                             
fatfs: done: /clients/common/dosemu2.git/test-imagedir/dXXXXs/c                 
~~~

I'll need to investigate further, as I just ran again there's no crash, but clearly there's something wrong as it doesn't move on after the breakpoint was hit.
~~~
dosdebug> bpintd 31
dosdebug> g
system state: stopped in DPMI-16bit

EAX: 00010000 EBX: 000200e6 ECX: 00010001 EDX: 000007bf eflags: 00013246
ESI: 00000772 EDI: 00000082 EBP: 0000091c DS: 00af ES: 00af FS: 0000 GS: 0000
 CS:EIP= 00a7:0000022b SS:ESP= 00af:00000760

#00a7:022b CD31             int  31

dosdebug> u #00a7:022b

#00a7:022b CD31             int  31
#00a7:022d 7207             jc   0236 ($+7)
#00a7:022f A31406           mov  [0614],ax
#00a7:0232 31C0             xor  ax,ax
#00a7:0234 CD31             int  31
#00a7:0236 0F82F301         jb   042D ($+1f3)
#00a7:023a A31606           mov  [0616],ax
#00a7:023d 668B0E1C00       mov  ecx,[001C]
#00a7:0242 B80105           mov  ax,0501
#00a7:0245 8B1E1E00         mov  bx,[001E]
#00a7:0249 CD31             int  31

dosdebug> t
system state: stopped in DPMI-16bit

EAX: 000100f7 EBX: 000200e6 ECX: 00010001 EDX: 000007bf eflags: 00003246
ESI: 00000772 EDI: 00000082 EBP: 0000091c DS: 00af ES: 00af FS: 0000 GS: 0000
 CS:EIP= 00a7:0000022d SS:ESP= 00af:00000760

#00a7:022d 7207             jc   0236 ($+7)

dosdebug> t
system state: stopped in DPMI-16bit

EAX: 000100f7 EBX: 000200e6 ECX: 00010001 EDX: 000007bf eflags: 00003346
ESI: 00000772 EDI: 00000082 EBP: 0000091c DS: 00af ES: 00af FS: 0000 GS: 0000
 CS:EIP= 00a7:0000022d SS:ESP= 00af:00000760

#00a7:022d 7207             jc   0236 ($+7)

~~~

stsp commented:
You know whats best, but please
do what your time permits. :)

andrewbird commented:
I tweaked the patch a little by saving the bpload index, seems to tidy up okay on program exit.

andrewbird commented:
After seeing what I thought was odd behaviour on my branch I just checked bpload on devel and I'm seeing the same thing. Can you explain what's going on here, please?

~~~
dosdebug> bpload
dosdebug> g
bpload: intercepting EXEC
INT 0x21, system state: running
AX=4b00  BX=050f  CX=03e9  DX=0535  SI=0100  DI=0100  SP=060a  BP=0089
DS=03e9  ES=03e9  FS=0000  GS=0000  FL=000b0246
CS:IP=03e9:02b1       SS:SP=03e9:060a

03e9:02b1 FA               cli
bpload: INT3 caught at 3e9:2b1
system state: stopped
AX=0000  BX=0000  CX=0000  DX=0000  SI=0000  DI=0000  SP=0760  BP=0000
DS=04af  ES=04af  FS=0000  GS=0000  FL=000a0346
CS:IP=04bf:0054       SS:SP=04bf:0760

04bf:0054 0E               push cs
~~~

I see `INT3 caught at 3e9:2b1` and I presume that's just a different seg:ofs representation of somewhere near the address we seem to land on. The mcbs command shows:
~~~
dosdebug> mcbs

ADDR(LOW) PARAS  OWNER
0292:0000 0x0150 [DOS]
  => ADDR      PARAS TYPE USAGE
     0293:0000 0x000a [D] Driver (EMUFS   )
     029e:0000 0x0082 [F] Files
     0321:0000 0x0010 [X] FCBs Extension
     0332:0000 0x0020 [B] Buffers
     0353:0000 0x008f [L] CDS Array
03e3:0000 0x0004 [DOS]
03e8:0000 0x00a5 [COMMAND]
048e:0000 0x0004 [FREE]
0493:0000 0x0010 [ allowed]
04a4:0000 0x0009 [y contro]
04ae:0000 0x9b50 [MFSREAD] (END)

ADDR(UMA) PARAS  OWNER
9fff:0000 ------ [LINK]
c300:0000 0x003c [DOS]
  => ADDR      PARAS TYPE USAGE
     c301:0000 0x0009 [D] Driver (EMS     )
     c30b:0000 0x0031 [D] Driver (CDROM   )
c33d:0000 0x1cc1 [FREE]
dfff:0000 ------ [LINK]
f000:0000 0x0bff [FREE] (END)
~~~

So if I move over the memory prefix I see this which I think looks like a PSP.

~~~
dosdebug> u 04af:0000

04af:0000 CD20             int  20
04af:0002 FF9F009A         call far word [bx-6600]
04af:0006 F0               db F0
04af:0007 FE               db FE
04af:0008 1DF0B1           sbb  ax,B1F0
04af:000b 02E9             add  ch,cl
04af:000d 034A01           add  cx,[bp+si+01]
04af:0010 E90355           jmp  5516 ($+5503)
04af:0013 01E9             add  cx,bp
04af:0015 03E9             add  bp,cx
04af:0017 0301             add  ax,[bx+di]
04af:0019 0101             add  [bx+di],ax
04af:001b 0002             add  [bp+si],al
04af:001d FF               db FF
04af:001e FF               db FF
04af:001f FF               db FF
~~~

But I can't figure out why we end up where we do, can you?

stsp commented:
Try loading com file, and make sure no
shell like DN/NC is working.
exe files may have entry points anywhere,
but with com you'll see if things are right
or wrong.

andrewbird commented:
By running the FOSSIL .com program things look more expected, except for a couple of odd memory blocks displayed (MS-DOS 6.22) in mcbs.
~~~
dosdebug> bpload
dosdebug> g
bpload: intercepting EXEC
INT 0x21, system state: running
AX=4b00  BX=050f  CX=03e9  DX=0535  SI=0100  DI=0100  SP=060a  BP=0081
DS=03e9  ES=03e9  FS=0000  GS=0000  FL=000b0246
CS:IP=03e9:02b1       SS:SP=03e9:060a

03e9:02b1 FA               cli
bpload: INT3 caught at 3e9:2b1
system state: stopped
AX=0000  BX=0000  CX=0000  DX=0000  SI=0000  DI=0000  SP=fffe  BP=0000
DS=04af  ES=04af  FS=0000  GS=0000  FL=000a0346
CS:IP=04af:0100       SS:SP=04af:fffe

04af:0100 FC               cld

dosdebug> mcbs

ADDR(LOW) PARAS  OWNER
0292:0000 0x0150 [DOS]
  => ADDR      PARAS TYPE USAGE
     0293:0000 0x000a [D] Driver (EMUFS   )
     029e:0000 0x0082 [F] Files
     0321:0000 0x0010 [X] FCBs Extension
     0332:0000 0x0020 [B] Buffers
     0353:0000 0x008f [L] CDS Array
03e3:0000 0x0004 [DOS]
03e8:0000 0x00a5 [COMMAND]
048e:0000 0x0004 [FREE]
0493:0000 0x0010 [ allowed]
04a4:0000 0x0009 [y contro]
04ae:0000 0x9b50 [FOSSIL] (END)

ADDR(UMA) PARAS  OWNER
9fff:0000 ------ [LINK]
c300:0000 0x003c [DOS]
  => ADDR      PARAS TYPE USAGE
     c301:0000 0x0009 [D] Driver (EMS     )
     c30b:0000 0x0031 [D] Driver (CDROM   )
c33d:0000 0x1cc1 [FREE]
dfff:0000 ------ [LINK]
f000:0000 0x0bff [FREE] (END)
~~~

So I don't think there's anything wrong here, just my expectations!

stsp commented:
No sure.
I am not getting this extra `03e9:02b1 FA               cli`
so there may still be something wrong.

andrewbird commented:
My test above is with MS-DOS 6.22 on devel, so perhaps it doesn't quite work correctly with that. I'll switch to FreeDOS 1.20 to see if things change.

stsp commented:
With freecom getting this:
```
bpload: intercepting EXEC
INT 0x21, system state: running
AX=4b00  BX=0896  CX=c8c1  DX=0716  SI=08b4  DI=0000  SP=0896  BP=3766
DS=c8c1  ES=c8c1  FS=0000  GS=0000  FL=000b0296
CS:IP=c8c1:095d       SS:SP=c8c1:0896

c8c1:095d 7202             jc   0961 ($+2)
bpload: INT3 caught at c8c1:95d
system state: stopped
AX=0000  BX=0000  CX=0000  DX=0000  SI=0000  DI=0000  SP=fffe  BP=0000
DS=0302  ES=0302  FS=0000  GS=0000  FL=000a0346
CS:IP=0302:0100       SS:SP=0302:fffe

0302:0100 FC               cld
```

stsp commented:
So somehow freecom gets in the way.

stsp commented:
Actually no, its just that with comcom32
there is an extra hlt. I'll amend this.

andrewbird commented:
I'll hang on for your changes then rebase my stuff.

stsp commented:
All done, output should now be better.
